Add slice type to `TypeName`

nathanwhit opened the pull request:
Closes #440.

nathanwhit commented:
> Can you add explicit tests that slices are _not_ copy or clone?

Done!

jackh726 commented:
@nathanwhit can you rebase. And then also update the well known types table for refs since I didn't think about it before I merged your other PR 🤦‍♀️

nathanwhit commented:
> @nathanwhit can you rebase? And then also update the well known types table for refs since I didn't think about it before I merged your other PR 🤦‍♀️

Rebased and updated the table! Should be good to go now.
